File: flutter_sdk/lib/src/transport/rt_event.dart
import 'dart:convert';
import 'dart:typed_data';

/// Sealed class representing all real-time WebSocket event types.
sealed class RtEvent {
  const RtEvent();

  /// Deserializes a JSON frame from the WebSocket into an [RtEvent].
  /// Returns null for unknown frame types (forward compatibility).
  static RtEvent? fromJson(Map<String, dynamic> json) {
    final type = json['type'] as String?;
    switch (type) {
      case 'message':
        return MessageRtEvent.fromJson(json);
      case 'low_otpk':
        return LowOtpkRtEvent.fromJson(json);
      case 'member_added':
        return MemberAddedRtEvent.fromJson(json);
      case 'member_removed':
        return MemberRemovedRtEvent.fromJson(json);
      case 'sender_key_distribution':
        return SenderKeyDistributionRtEvent.fromJson(json);
      case 'ping':
        return const PingRtEvent();
      default:
        return null; // Unknown type — ignore for forward compatibility
    }
  }
}

/// A new encrypted message in a conversation.
class MessageRtEvent extends RtEvent {
  final String conversationId;
  final int seq;
  final String senderUserId;
  final String senderDeviceId;
  final Uint8List ciphertext;
  final Map<String, dynamic> protocolHeader;
  final int serverTs;
  final String? attachmentId;

  const MessageRtEvent({
    required this.conversationId,
    required this.seq,
    required this.senderUserId,
    required this.senderDeviceId,
    required this.ciphertext,
    required this.protocolHeader,
    required this.serverTs,
    this.attachmentId,
  });

  factory MessageRtEvent.fromJson(Map<String, dynamic> json) =>
      MessageRtEvent(
        conversationId: json['conversation_id'] as String,
        seq: json['seq'] as int,
        senderUserId: json['sender_user_id'] as String,
        senderDeviceId: json['sender_device_id'] as String,
        ciphertext: Uint8List.fromList(
            base64.decode(json['ciphertext'] as String)),
        protocolHeader:
            json['protocol_header'] as Map<String, dynamic>? ?? {},
        serverTs: json['server_ts'] as int,
        attachmentId: json['attachment_id'] as String?,
      );
}

/// The server is running low on one-time pre-keys for this device.
class LowOtpkRtEvent extends RtEvent {
  final String deviceId;
  final int count;

  const LowOtpkRtEvent({required this.deviceId, required this.count});

  factory LowOtpkRtEvent.fromJson(Map<String, dynamic> json) =>
      LowOtpkRtEvent(
        deviceId: json['device_id'] as String,
        count: json['count'] as int,
      );
}

/// A new member has been added to a group conversation.
class MemberAddedRtEvent extends RtEvent {
  final String conversationId;
  final String userId;
  final List<String> devices;

  const MemberAddedRtEvent({
    required this.conversationId,
    required this.userId,
    required this.devices,
  });

  factory MemberAddedRtEvent.fromJson(Map<String, dynamic> json) =>
      MemberAddedRtEvent(
        conversationId: json['conversation_id'] as String,
        userId: json['user_id'] as String,
        devices: List<String>.from(json['devices'] as List? ?? []),
      );
}

/// A member has been removed from a group conversation.
class MemberRemovedRtEvent extends RtEvent {
  final String conversationId;
  final String userId;

  const MemberRemovedRtEvent({
    required this.conversationId,
    required this.userId,
  });

  factory MemberRemovedRtEvent.fromJson(Map<String, dynamic> json) =>
      MemberRemovedRtEvent(
        conversationId: json['conversation_id'] as String,
        userId: json['user_id'] as String,
      );
}

/// A sender key distribution message for a group conversation.
class SenderKeyDistributionRtEvent extends RtEvent {
  final String conversationId;
  final String senderUserId;
  final Uint8List encryptedSkdm;

  const SenderKeyDistributionRtEvent({
    required this.conversationId,
    required this.senderUserId,
    required this.encryptedSkdm,
  });

  factory SenderKeyDistributionRtEvent.fromJson(Map<String, dynamic> json) =>
      SenderKeyDistributionRtEvent(
        conversationId: json['conversation_id'] as String,
        senderUserId: json['sender_user_id'] as String,
        encryptedSkdm: Uint8List.fromList(
            base64.decode(json['encrypted_skdm'] as String)),
      );
}

/// Server ping — the SDK responds with pong automatically.
class PingRtEvent extends RtEvent {
  const PingRtEvent();
}

File: web_sdk/src/transport/rt-event.ts
// src/transport/rt-event.ts

import { base64urlDecode } from '../utils/encoding.js';

export type ProtocolHeader =
  | {
      type: 'x3dh_init';
      ek: Uint8Array;
      spkId: number;
      otpkId?: number;
    }
  | {
      type: 'double_ratchet';
      dh: Uint8Array;
      n: number;
      pn: number;
    }
  | {
      type: 'sender_key';
      chainId: number;
      iteration: number;
    };

export type RtEvent =
  | {
      type: 'message';
      conversationId: string;
      seq: number;
      senderUserId: string;
      senderDeviceId: string;
      ciphertext: Uint8Array;
      protocolHeader: ProtocolHeader;
      serverTs: number;
      attachmentId: string | null;
    }
  | {
      type: 'low_otpk';
      deviceId: string;
      count: number;
    }
  | {
      type: 'member_added';
      conversationId: string;
      userId: string;
      devices: string[];
    }
  | {
      type: 'member_removed';
      conversationId: string;
      userId: string;
    }
  | {
      type: 'sender_key_distribution';
      conversationId: string;
      senderUserId: string;
      encryptedSkdm: Uint8Array;
    }
  | {
      type: 'ping';
    };

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function parseRtEvent(raw: any): RtEvent | null {
  if (!raw || typeof raw.type !== 'string') return null;

  switch (raw.type) {
    case 'message': {
      if (!raw.envelope) return null;
      const header = parseProtocolHeader(raw.envelope.protocol_header);
      if (!header) return null;
      return {
        type: 'message',
        conversationId: String(raw.conversation_id ?? raw.envelope.conversation_id),
        seq: Number(raw.seq),
        senderUserId: String(raw.sender_user_id),
        senderDeviceId: String(raw.sender_device_id),
        ciphertext: base64urlDecode(String(raw.envelope.ciphertext)),
        protocolHeader: header,
        serverTs: Number(raw.server_ts),
        attachmentId: raw.envelope.attachment_id ?? null,
      };
    }
    case 'low_otpk':
      return {
        type: 'low_otpk',
        deviceId: String(raw.device_id),
        count: Number(raw.count),
      };
    case 'member_added':
      return {
        type: 'member_added',
        conversationId: String(raw.conversation_id),
        userId: String(raw.user_id),
        devices: Array.isArray(raw.devices) ? raw.devices.map(String) : [],
      };
    case 'member_removed':
      return {
        type: 'member_removed',
        conversationId: String(raw.conversation_id),
        userId: String(raw.user_id),
      };
    case 'sender_key_distribution':
      return {
        type: 'sender_key_distribution',
        conversationId: String(raw.conversation_id),
        senderUserId: String(raw.sender_user_id),
        encryptedSkdm: base64urlDecode(String(raw.encrypted_skdm)),
      };
    case 'ping':
      return { type: 'ping' };
    default:
      // Unknown event types silently discarded for forward compatibility
      return null;
  }
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function parseProtocolHeader(h: any): ProtocolHeader | null {
  if (!h || typeof h.type !== 'string') return null;
  switch (h.type) {
    case 'x3dh_init':
      return {
        type: 'x3dh_init',
        ek: base64urlDecode(String(h.ek)),
        spkId: Number(h.spk_id),
        ...(h.otpk_id !== undefined ? { otpkId: Number(h.otpk_id) } : {}),
      } as ProtocolHeader;
    case 'double_ratchet':
      return {
        type: 'double_ratchet',
        dh: base64urlDecode(String(h.dh)),
        n: Number(h.n),
        pn: Number(h.pn),
      };
    case 'sender_key':
      return {
        type: 'sender_key',
        chainId: Number(h.chain_id),
        iteration: Number(h.iteration),
      };
    default:
      return null;
  }
}
